### PR TITLE
test(repository-tests): adding acceptance tests for hasManyThrough relation

### PR DESCRIPTION
**Pull request by agnes512**
#4438 #2359

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [ ] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [ ] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

File: packages/repository-tests/src/crud/relations/acceptance/has-many-through.relation.acceptance.ts [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+    it('can patch an instance', async () => {
+      const item1 = await customerRepo
+        .cartItems(existingCustomerId)
+        .create({description: 'group 1'});
+      const item2 = await customerRepo
+        .cartItems(existingCustomerId)
+        .create({description: 'group 1'});
+      const count = await customerRepo
+        .cartItems(existingCustomerId)
+        .patch({description: 'group 2'}, {id: item2.id});
```
Comment (jannyHou) [inline packages/repository-tests/src/crud/relations/acceptance/has-many-through.relation.acceptance.ts:120]: nitpick: If filter `{id: item2.id}` not provided, would it patch all items belong to customer? If so maybe add one more test for patch all with no filter.

File: packages/repository-tests/src/crud/relations/acceptance/has-many-through.relation.acceptance.ts [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+    it('throws error when query tries to change the target id', async () => {
+      // a diff id for CartItem instance
+      const anotherId = (await givenPersistedCustomerInstance()).id;
+      await customerRepo
+        .cartItems(existingCustomerId)
+        .create({description: 'group 1'});
+
+      await expect(
+        customerRepo.cartItems(existingCustomerId).patch({id: anotherId}),
+      ).to.be.rejectedWith(/Property "id" cannot be changed!/);
```
Comment (jannyHou) [inline packages/repository-tests/src/crud/relations/acceptance/has-many-through.relation.acceptance.ts:141]: 👍

File: packages/repository-tests/src/crud/relations/acceptance/has-many-through.relation.acceptance.ts [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+      await customerRepo
+        .cartItems(existingCustomerId)
+        .create({description: 'group 1'});
+
+      let links = await customerCartItemLinkRepo.find();
+      let cartItems = await cartItemRepo.find();
+      expect(links).have.length(2);
+      expect(cartItems).have.length(2);
+
+      await customerRepo.cartItems(existingCustomerId).delete();
```
Comment (jannyHou) [inline packages/repository-tests/src/crud/relations/acceptance/has-many-through.relation.acceptance.ts:157]: nitpick: does `delete` takes in a filter? Maybe add another test for delete with filter

File: packages/repository-tests/src/crud/relations/acceptance/has-many-through.relation.acceptance.ts [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+      cartItems = await cartItemRepo.find();
+      expect(links).have.length(0);
+      expect(cartItems).have.length(0);
+    });
+
+    it('can link a target model to a source model', async () => {
+      const item = await cartItemRepo.create({description: 'an item'});
+
+      let targets = await customerRepo.cartItems(existingCustomerId).find();
+      expect(targets).to.deepEqual([]);
```
Comment (jannyHou) [inline packages/repository-tests/src/crud/relations/acceptance/has-many-through.relation.acceptance.ts:191]: nitpick: `to.be.empty()`

File: packages/repository-tests/src/crud/relations/acceptance/has-many-through.relation.acceptance.ts (hunk 3 above)
Comment (agnes512) [reply packages/repository-tests/src/crud/relations/acceptance/has-many-through.relation.acceptance.ts:157]: Just realized that the `delete` method has a bug. It is able to delete **targets** based on the filter but it deletes **all through instance that has the souceKey**. Will open up an issue to fix it.

Comment (agnes512) [reply packages/repository-tests/src/crud/relations/acceptance/has-many-through.relation.acceptance.ts:178]: The issue is being tracked in https://github.com/strongloop/loopback-next/issues/5852

**Pull request merged**